### PR TITLE
add `MP_SMALL_STACK_SIZE` option

### PR DESCRIPTION
**Pull request by sjaeckel**
This adds an option to use a heap-buffer for the usually stack-based
`MP_WARRAY`-sized temporary buffers.

Per default it will reserve a single buffer, which can be modified
* at compile-time via the `MP_WARRAY_NUM` define
* at run-time by calling `mp_warray_init()`

The internal structure can only be created once. If one wants to modify
the maximum number of elements, the entire structure has to be free'd
by calling `mp_warray_free()`.

In case one wants to use this option with multiple threads, one shall
use the `mp_warray_init()` function and pass appropriate locking functions.

### Timings

| op / algo | ECC | RSA |
| ------------ | ------ | ------ |
| decrypt_key | ecc_decrypt_key | rsa_decrypt_key |
| encrypt_key | ecc_encrypt_key | rsa_encrypt_key |
| sign_hash | ecc_sign_hash | rsa_sign_hash |
| verify_hash | ecc_verify_hash | rsa_verify_hash |

Related to: #441 #447 #511

### CODE REVIEW

**Comment by sjaeckel** [comment]
@jlaitine could you maybe have a look and tell me whether that's an acceptable solution for you?

File: doc/bn.tex [hunk 1]
```diff
@@ -428,7 +442,9 @@ \chapter{Getting Started with LibTomMath}
 \section{Building Programs}
 In order to use LibTomMath you must include ``tommath.h'' and link against the appropriate library
 file (typically
-libtommath.a).	There is no library initialization required and the entire library is thread safe.
+libtommath.a).	There is no library initialization required and the entire library is thread safe
+if it is used in its default configuration. Locking is recommended if the small-stack option
+is enabeled and multiple threads are used, c.f. \ref{ch:SMALL_STACK_INTRO} resp. \ref{ch:SMALL_STACK_API}
```
Comment (czurnieden) [inline doc/bn.tex:447]: s/enabeled/enabled/

File: doc/bn.tex [hunk 2]
```diff
@@ -813,6 +829,53 @@ \subsection{Adding additional digits}
...
+In case the \texttt{MP\_SMALL\_STACK\_SIZE} symbol is defined the following functions
+can be useful.
+
+To initialize the internal structure the following function shall be called.
+
+\index{mp\_warray\_init}
+\begin{alltt}
+mp_err mp_warray_init(size_t n_alloc, bool preallocate, mp_lock *lock);
+\end{alltt}
+
```
Comment (czurnieden) [inline doc/bn.tex:844]: Maybe add a sentence about the flag `preallocate`? It is clear in the code but a bit ambiguous here.

**Review by czurnieden — CHANGES_REQUESTED** [review]
Found nothing except some typos.

But the effect on the ECC computations *is* interesting!

**Comment by czurnieden** [comment]
What happened to this PR? Did you change your mind? I find it quite interesting to have a choice between stack and heap for this. On the other side: it breaks our general thread-safety and needs locks.
But I'm still supportive.

File: s_mp_warray_get.c [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+      if (s_mp_warray.l_free[n].warray) {
+         s_mp_warray.l_used[n] = s_mp_warray.l_free[n];
+         s_mp_warray.l_free[n].warray = NULL;
+         ret = s_mp_warray.l_used[n].warray;
+         goto LBL_OUT;
+      }
+   }
+   if (s_mp_warray.allocated + 1 > s_mp_warray.usable)
+      goto LBL_OUT;
+   ret = MP_CALLOC(MP_WARRAY, sizeof(mp_word));
```
[the remark this answers is not in the document]
Comment (sjaeckel) [reply s_mp_warray_get.c:25]: I guess we have to live with the worst-case assumption as the default value.

File: .github/workflows/main.yml [hunk 4]
```diff
@@ -153,7 +159,7 @@ jobs:
       matrix:
         os: [ ubuntu-18.04, ubuntu-20.04 ]
         build_type: [ '', -DCMAKE_BUILD_TYPE=Debug, -DCMAKE_BUILD_TYPE=Release, -DCMAKE_BUILD_TYPE=RelWithDebInfo, -DCMAKE_BUILD_TYPE=MinSizeRel ]
-        cc: [ clang, gcc ]
+        cc: [ clang-10, gcc ]
```
[the remark this answers is not in the document]
Comment (sjaeckel) [reply .github/workflows/main.yml]: I can't really remember, but maybe it had to do with the CI runners at the time of the original PR still being 18.04 and 20.04, where 18.04 default clang was too old or something like that.

Fixed.

**Comment by sjaeckel** [comment]
> What happened to this PR? Did you change your mind?

No, I just didn't have the passion to follow up :)

> On the other side: it breaks our general thread-safety and needs locks.

That's indeed a thing. I'm not sure whether it'd make sense to replace the locks by taking care of that ourselves and using lock-free/atomic operations under the hood!?

**Comment by sjaeckel** [comment]
> I'm not sure whether it'd make sense to replace the locks by taking care of that ourselves and using lock-free/atomic operations under the hood!?

What do you think? Better like that!?

**Review by czurnieden — APPROVED** [review]

**Comment by czurnieden** [comment]
Sorry, took a bit longer to check.

> Better like that!?

Yes, looks reasonable.

Caveat: I cannot test the Windows stuff now: my last PC here with Windows (what is the plural of Windows?) let the magic smoke out, need a visit to Ebay before going on.

PS: I am aware of my TODO list, it's not forgotten!

**Comment by sjaeckel** [comment]
> what is the plural of Windows?

My SO just explained to me that the plural only works with the complete name! Microsoft Windows -> Microsoft**s** Windows

> my TODO list

No need to hurry, we have time.

> test the Windows stuff

I'll see what I can do by extending CI.

**Comment by czurnieden** [comment]
> My SO just explained to me that the plural only works with the complete name! Microsoft Windows -> Microsofts Windows

Ah, ok, my thanks to your SO!

> I'll see what I can do by extending CI.

That would be really nice!

> No need to hurry, we have time.

You have, but I'm an overweight old fart already! ;-)

My main problem is the error I got with the base2->base10 conversion with a single round of NR to compute the next divisor at a very large (close to `MP_MAX_DIGIT_COUNT` with `MP_64BIT` ) number. Sadly a randomly generated large number and it hasn't been logged what it was (Full HD. Yes, I'm an idiot). Haven't been able to repeat the error and am trying to compute the accumulated error of the NR rounds now manually. Fiddly.

We need to be able to do at most 22 rounds with `n_0 = 1000` and the individual absolute error is about `2^(-34)`. But it is not floating point, it is integer math, so what is the accumulated error at the end? Does it even accumulate?

Was the aformentioned single error just a random bitflip? An error elsewhere?

\*hng\*!

**Comment by sjaeckel** [comment]
> Ah, ok, my thanks to your SO!

You're very welcome ;-)

> > I'll see what I can do by extending CI.

wow, what a ride ... In between I was not sure whether I'd simply disable the feature for MSVC ... but now it seems to work^TM ...

**Comment by sjaeckel** [comment]
@jlaitine can you please check if this works for you?

**Comment by sjaeckel** [comment]
> In between I was not sure whether I'd simply disable the feature for MSVC ... but now it seems to work^TM ...

aaaand in the end I disabled it for MSVC since IMO the most elegant solution is via TLS and MSVC seems to have a problem with TLS in DLLs. I'm not 100% sure what's going wrong, but since it was going fine with the static library [0] I suspect that we'd have to implement what's described in [1] and I'm simply not willing to do that, nor would I accept a PR that implements it.

If I'm mistaken there and someone with more MSVC experience wants to chime in and provide an easy fix, I'll happily accept that :)

[0] https://ci.appveyor.com/project/libtom/libtommath/builds/49507598
[1] https://learn.microsoft.com/en-us/windows/win32/dlls/using-thread-local-storage-in-a-dynamic-link-library

**Review by czurnieden — APPROVED** [review]
Bearhug?

**Comment by czurnieden** [comment]
> https://learn.microsoft.com/en-us/windows/win32/dlls/using-thread-local-storage-in-a-dynamic-link-library

So that's the superlativ of "unnecessary complicated", isn't it?

**Comment by jlaitine** [comment]
> @jlaitine can you please check if this works for you?

Hi sorry for the very long response time, I have been busy with some other things. I see that this already got merged, so I tested the development branch directly and this change works perfectly!

I still have a small compilation error, in PX4, which is compiled with "-Wall". Not related to this PR. I created anoter PR out of that: https://github.com/libtom/libtommath/pull/578

Thanks again,
Jukka

**Comment by sjaeckel** [comment]
Thanks a lot for the feedback! Glad to hear that it works for you :)

**Pull request merged**